### PR TITLE
Chat - implement dayHeaderFormat and messageTimestampFormat options

### PR DESCRIPTION
**Pull request by Zedwag**

### CODE REVIEW

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/chat.ts]: JFYI: you will have conflict when https://github.com/DevExpress/DevExtreme/pull/28219 is merged

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/chat.ts]: https://github.com/Zedwag/DevExtreme/blob/9614145c4838d3944504d7fb2af44433f4d1d4e4/packages/devextreme/js/__internal/ui/date_box/m_date_box.base.ts#L80-L81

dateSerializationFormat option has `undefined` default value, why we should use null? If we will use `undefined` it should looks like
```
messageTimestampFormat?: Format;
```

Comment (EugeniyKiyashko) [reply packages/devextreme/js/__internal/ui/chat/chat.ts]: Here we should have the same types that are added in correspond d.ts file

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/chat.ts]: How do you think if we add line-break after each of options it will looks better?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/chat.tests.js
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/chat.tests.js]:
```suggestion
        QUnit.test('dayHeaderFormat option value should be passed to messageList on init', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js:398]: I understand that you calculate the expected value instead of using a static one because you get new Date() values every day. However, it seems better to mock new Date() using sinon.useFakeTimers() as following
```
const clock = sinon.useFakeTimers(new Date('2021/10/17'));
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js]: I understand that you calculate the expected value instead of using a static one because you get new Date() values every day. However, it seems better to mock new Date() using sinon.useFakeTimers() as following
```
const clock = sinon.useFakeTimers(new Date('2021/10/17'));
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js]: Runtime change tests seem to be missing

File: packages/devextreme/js/__internal/ui/chat/messagelist.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/messagelist.ts]: Why we not use one of predefined format 'shortdate' as default value. For example

```
     dateLocalization.format(date, messageTimestampFormat ?? 'shortdate');
```
https://github.com/Zedwag/DevExtreme/blob/9614145c4838d3944504d7fb2af44433f4d1d4e4/packages/devextreme/js/localization/date.js#L19

Instead of this we use `date.toLocaleDateString()`. Moreover we work with transformation of Date() twice if dayHeaderFormat is defined.

Well, toLocaleDateString
1) supports all browsers that we support
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString
2) returns a string with a language-sensitive representation it means that it shouldn't contain '-' that you replace on the line 221.

image
image

Please look into this matter in more detail to ensure that we are not making a mistake

File: packages/devextreme/js/__internal/ui/chat/messagegroup.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: Seem it relates with https://github.com/DevExpress/DevExtreme/pull/28234/files#r1813625995

File: packages/devextreme/js/__internal/ui/chat/messagegroup.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: Did we approve that we should use this format by default? I mean 'hour12: false' without using AM/PM. I didn't find any information in specification.

Comment (ksercs) [reply packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: Yeah, it looks strange at least because `hour12: false`, it will be not expected in EN locale. 
I believe it should be smth like "hh:mm" or like this (please find similar solution in DateBox).
And I believe we should always use `dateLocalization.format`, not `date.toLocaleTimeString`

File: packages/devextreme/js/__internal/ui/chat/messagelist.ts
Comment (ksercs) [reply packages/devextreme/js/__internal/ui/chat/messagelist.ts]: Yeah, i believe we should always use `dateLocalization.format` with some format.
Maybe we even should get rid of `null` in type and set some default format instead of this - I think it'll be much clearer

File: packages/devextreme/js/__internal/ui/chat/messagegroup.ts
Comment (EugeniyKiyashko) [reply packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: We were going to think about it on 24 July here: https://github.com/DevExpress/DevExtreme/pull/27815#discussion_r1689579976. But I don't see any cards that were created after the discussion.

Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: now it used `dateLocalization.format` in both cases. Default format is `HH:mm` ed2574f9d5f3c5a301c284eb8acaba16c73321ad

File: packages/devextreme/js/__internal/ui/chat/messagegroup.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: ed2574f9d5f3c5a301c284eb8acaba16c73321ad

File: packages/devextreme/js/__internal/ui/chat/messagelist.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/messagelist.ts]: ed2574f9d5f3c5a301c284eb8acaba16c73321ad

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/chat.tests.js
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/chat.tests.js]: ed2574f9d5f3c5a301c284eb8acaba16c73321ad

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/chat.ts]: ed2574f9d5f3c5a301c284eb8acaba16c73321ad

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/chat.ts]: ed2574f9d5f3c5a301c284eb8acaba16c73321ad

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js]: added

**Review by ksercs — DISMISSED** [review]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js]: I left it as it is, since idk how to fix no clear timeout error when i use this approach (clock.restore() doesn't help)

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/chat.ts]: It's better to consider that these properties are optional as it will in public d.ts file

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/chat.ts:65]: Just for clarification: Did you approve option names and default values with PM & techwriter?

File: packages/devextreme/js/__internal/ui/chat/messagelist.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/messagelist.ts]: Does it require to add '?? 'shorttime''? The default value of messageTimestampFormat is already 'shorttime' or did I miss something?

File: packages/devextreme/js/__internal/ui/chat/messagelist.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/messagelist.ts]:
```suggestion
      .text(headerDate)
```

File: packages/devextreme/js/__internal/ui/chat/messagegroup.ts
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/chat/messagegroup.ts]:
```suggestion
    return formattedTime;
```

Comment (EugeniyKiyashko) [reply packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: It strange to see that you use templated literals for type conversion here. Let's improve it and choose other way to avoid ts issues

Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/messagegroup.ts]: https://github.com/DevExpress/DevExtreme/pull/28234/commits/146df9b953cfca2760258090346ce0e5f1173871

File: packages/devextreme/js/__internal/ui/chat/messagelist.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/messagelist.ts]: https://github.com/DevExpress/DevExtreme/pull/28234/commits/146df9b953cfca2760258090346ce0e5f1173871

File: packages/devextreme/js/__internal/ui/chat/messagelist.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/messagelist.ts]: https://github.com/DevExpress/DevExtreme/pull/28234/commits/146df9b953cfca2760258090346ce0e5f1173871

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/chat.ts:65]: yes, names were approved by pm and tw

File: packages/devextreme/js/__internal/ui/chat/chat.ts
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/chat/chat.ts]: https://github.com/DevExpress/DevExtreme/pull/28234/commits/146df9b953cfca2760258090346ce0e5f1173871

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/chatParts/messageList.tests.js]: https://github.com/DevExpress/DevExtreme/pull/28234/commits/146df9b953cfca2760258090346ce0e5f1173871

**Review by EugeniyKiyashko — DISMISSED** [review]

**Review by ksercs — DISMISSED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Comment by EugeniyKiyashko** [comment]
@ksercs There haven't been any changes since your last approval. Therefore, I won't wait for your approval and will merge.

**Pull request merged**
